### ISSUE TITLE
32-bit test_bitcoin failure: error in "MempoolAncestorIndexingTest"

### ISSUE THREAD

**Issue by error10** [bitcoinxt/bitcoinxt#125]
On 32-bit Fedora 22 I get the following test failures from test_bitcoin.  When compiled on 64-bit Fedora 22, all tests pass.

```
Running 191 test cases...
test/mempool_tests.cpp(112): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [c9f5883d7f47175a15328365533f0f501163bfd6886ed9b03bca5b7503c2f0c3 != c0890433e773ca3ab910a0541045115ab85dc849ea7a47571f229389992bf35d]
test/mempool_tests.cpp(112): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [2371dcf167f394888b4f70c609f048b1e4ed090bef36dbfd0e3248960da46820 != c9f5883d7f47175a15328365533f0f501163bfd6886ed9b03bca5b7503c2f0c3]
test/mempool_tests.cpp(112): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [5500cec6e62b19fbcc8f6367dd6188ffcf85086c5d6321a103345e9cf27e4fef != 2371dcf167f394888b4f70c609f048b1e4ed090bef36dbfd0e3248960da46820]
test/mempool_tests.cpp(112): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [940d520187fa88a9210b765398b31ecd5682e79ee1d93e887cef502195ffc30f != 5500cec6e62b19fbcc8f6367dd6188ffcf85086c5d6321a103345e9cf27e4fef]
test/mempool_tests.cpp(112): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [c0890433e773ca3ab910a0541045115ab85dc849ea7a47571f229389992bf35d != 940d520187fa88a9210b765398b31ecd5682e79ee1d93e887cef502195ffc30f]

*** 5 failures detected in test suite "Bitcoin Test Suite"
FAIL test/test_bitcoin (exit status: 201)
```

**Comment by dgenr8** [comment]
Could you add some logging or otherwise get a stack trace revealing which call to CheckSort<3> is causing this?

**Comment by error10** [comment]
@dgenr8 After sprinkling `BOOST_TEST_MESSAGE()` liberally, I found that the call generating the five test failures is on line 376, the final call in the function.

```
    sortedOrder.erase(sortedOrder.begin()+1);
    sortedOrder.pop_back();
    sortedOrder.insert(sortedOrder.begin(), tx7.GetHash().ToString());
    CheckSort<3>(pool, sortedOrder);
}

BOOST_AUTO_TEST_SUITE_END()
```

**Comment by dgenr8** [comment]
@error10 Could you add the line marked with + in the CheckSort function

```
    for (; it != pool.mapTx.get<indexNumber>().end(); ++it, ++count) {
+    	BOOST_TEST_MESSAGE("fee: " << it->GetFee() << " fee+: " << it->GetFeesWithAncestors() << " size+: " << it->GetSizeWithAncestors());
        BOOST_CHECK_EQUAL(it->GetTx().GetHash().ToString(), sortedOrder[count]);
```

...and post the output from the last two tests under 32-bit, running

``src/test/test_bitcoin --run_test=mempool_tests --log_level=all``

**Comment by error10** [comment]
@dgenr8 Here's the last bit of the test run. I left a bit more in than you asked for, for some additional context.

```
test/mempool_tests.cpp(364): info: check pool.size() == 7 passed
test/mempool_tests.cpp(108): info: check pool.size() == sortedOrder.size() passed
fee: 20000 fee+: 20000 size+: 21
test/mempool_tests.cpp(113): info: check it->GetTx().GetHash().ToString() == sortedOrder[count] passed
fee: 79967 fee+: 79967 size+: 84
test/mempool_tests.cpp(113): info: check it->GetTx().GetHash().ToString() == sortedOrder[count] passed
fee: 15000 fee+: 15000 size+: 21
test/mempool_tests.cpp(113): info: check it->GetTx().GetHash().ToString() == sortedOrder[count] passed
fee: 10000 fee+: 10000 size+: 21
test/mempool_tests.cpp(113): info: check it->GetTx().GetHash().ToString() == sortedOrder[count] passed
fee: 10000 fee+: 10000 size+: 21
test/mempool_tests.cpp(113): info: check it->GetTx().GetHash().ToString() == sortedOrder[count] passed
fee: 0 fee+: 0 size+: 21
test/mempool_tests.cpp(113): info: check it->GetTx().GetHash().ToString() == sortedOrder[count] passed
fee: 0 fee+: 0 size+: 21
test/mempool_tests.cpp(113): info: check it->GetTx().GetHash().ToString() == sortedOrder[count] passed
test/mempool_tests.cpp(108): info: check pool.size() == sortedOrder.size() passed
fee: 20000 fee+: 20000 size+: 21
test/mempool_tests.cpp(113): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [c9f5883d7f47175a15328365533f0f501163bfd6886ed9b03bca5b7503c2f0c3 != c0890433e773ca3ab910a0541045115ab85dc849ea7a47571f229389992bf35d]
fee: 15000 fee+: 15000 size+: 21
test/mempool_tests.cpp(113): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [2371dcf167f394888b4f70c609f048b1e4ed090bef36dbfd0e3248960da46820 != c9f5883d7f47175a15328365533f0f501163bfd6886ed9b03bca5b7503c2f0c3]
fee: 10000 fee+: 10000 size+: 21
test/mempool_tests.cpp(113): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [5500cec6e62b19fbcc8f6367dd6188ffcf85086c5d6321a103345e9cf27e4fef != 2371dcf167f394888b4f70c609f048b1e4ed090bef36dbfd0e3248960da46820]
fee: 10000 fee+: 10000 size+: 21
test/mempool_tests.cpp(113): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [940d520187fa88a9210b765398b31ecd5682e79ee1d93e887cef502195ffc30f != 5500cec6e62b19fbcc8f6367dd6188ffcf85086c5d6321a103345e9cf27e4fef]
fee: 79967 fee+: 79967 size+: 4294967359
test/mempool_tests.cpp(113): error in "MempoolAncestorIndexingTest": check it->GetTx().GetHash().ToString() == sortedOrder[count] failed [c0890433e773ca3ab910a0541045115ab85dc849ea7a47571f229389992bf35d != 940d520187fa88a9210b765398b31ecd5682e79ee1d93e887cef502195ffc30f]
fee: 0 fee+: 0 size+: 21
test/mempool_tests.cpp(113): info: check it->GetTx().GetHash().ToString() == sortedOrder[count] passed
Leaving test case "MempoolAncestorIndexingTest"; testing time: 62520mks
Leaving test suite "mempool_tests"
Leaving test suite "Bitcoin Test Suite"

*** 5 failures detected in test suite "Bitcoin Test Suite"
```

**Comment by dgenr8** [comment]
@error10 Thank you.  Can you try modifying [this line](https://github.com/bitcoinxt/bitcoinxt/blob/master/src/txmempool.cpp#L266) to:

``int64_t modifySize = -((int64_t)removeIt->GetTxSize());``

**Comment by error10** [comment]
With this change, all tests pass.

**Comment by dgenr8** [comment]
@error10 Fixed in #127.
